clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -211,6 +211,7 @@ public String protocolType() {
     protected JoinGroupRequestData.JoinGroupRequestProtocolCollection metadata() {
         log.debug("Joining group with current subscription: {}", subscriptions.subscription());
         this.joinedSubscription = subscriptions.subscription();
+        subscriptions.resetGroupSubscription();
```

Comment:
I'm confused about a couple things. First, if a subscription changes while we are in the middle of a rebalance, do we need to call `onJoinPrepare` again? With the old semantics, we probably didn't need to because we revoked everything, but with "cooperative" rebalancing, perhaps we should?  cc @guozhangwang 

The second thing is that it seems surprising that we raise auth errors from the "group subscription." If consumer 1 has subscribed to "A" and consumer 2 has subscribed to "B," I probably wouldn't expect to get any auth errors for topic "B" from consumer 1. So maybe we should be ignoring auth errors not part of `subscription()`? Not sure if that is straightforward or if there are any additional implications.

Reply:
@hachikuji Thanks for the review. If we decide that we always need to do `onJoinPrepare`, then we won't need this fix.

It is odd to propagate authorization failures from the group subscription since that feels more like an implementation detail. But at the same time, if we don't get metadata for some topics, is there any point in adding topics from group subscription to the ConsumerMetadata? We probably need to update javadoc for poll() if we don't change the current behaviour.

Reply:
Regarding the first question, if a subscription changes while we are in the middle of a rebalance we would call `onPartitionsRevoked` immediately on those still-assigned partitions that are not part of the new subscription any more, and for the returned assignment we check if they match our join-subscription and if not we trigger another rebalance (this is pre-429). So I think we still do not need to re-trigger `onJoinPrepare` when subscription changes in the middle.

Reply:
@rajinisivaram I think I understand the problem a little better now and I'm coming around to the approach in this PR. It does make sense to me to reset the group subscription at the time that we set `joinedSubscription` in `ConsumerCoordinator`. Then I guess my only question is whether we still need to reset it inside `onJoinPrepare` as well?

Another (possibly worse) idea I had is to change the implementation of `resetGroupSubscription`. Perhaps instead of setting it equal to the current local subscription, we should set it to null or empty. We could try to maintain a cleaner separation between the local and group subscriptions. We would then need to update `ConsumerMetadata` to work with both the local and group subscription.

Reply:
@hachikuji Thanks for the reviews. I pushed two commits. The first one just removes reset from `onJoinPrepare` since we don't need it any more. I also wanted to try out your suggestion of maintaining clean separation between local and group subscriptions. The last commit does this. Do you think it is worth making that change?